app/models/case_contact_report.rb

```diff
@@ -2,11 +2,26 @@ class CaseContactReport
   attr_reader :case_contacts
 
   def initialize(args = {})
-    @case_contacts = if args[:start_date] && args[:end_date]
-      CaseContact.occurred_between(args[:start_date], args[:end_date])
-    else
-      CaseContact.all
-    end
+    @case_contacts = filtered_case_contacts(args)
+  end
+
+  def filtered_case_contacts(args)
+    return CaseContact.all if args.empty?
+
+    contact = CaseContact
+    contact = contact.where(supervisors: args[:supervisor_ids]) if args.has_key?(:supervisor_ids) # supervisor, as an array
+    contact = contact.where(creator_id: args[:creator_id]) if args.has_key?(:creator_id) # volunteer
+    contact = contact.occurred_between(args[:start_date], args[:end_date]) if  args.has_key?(:start_date) && args.has_key?(:end_date)
+    contact = contact.where(contact_made: args[:contact_made]) if args.has_key?(:contact_made) # contact made, boolean
+    contact = contact.joins(:casa_case).where("casa_cases.transition_aged_youth in (?)", args[:has_transitioned]) if args.has_key?(:has_transitioned) # boolean, but also in the casa cases table
+    contact = contact.where(want_driving_reimbursement: args[:want_driving_reimbursement]) if args.has_key?(:want_driving_reimbursement)
+    # This filter is not working properly, commenting it out for now
+    # if args[:contact_types]
+    #   args[:contact_types].each do |contact_type|
+    #     contact = contact.where "contact_types @> ARRAY[?]::varchar[]", [contact_type]
+    #   end
+    # end
+    contact
```

Comment:
@duaimei it looks like this won't take into account multiple filters getting checked, will it?

If you need to account for multiple filters, then a couple different ways I have done this is through either chaining separate scopes or chaining scopes using `inject`.

With scopes:

```
# app/models/case_contact.rb

scope :supervisors, ->(supervisor_ids) { where(supervisors: supervisor_ids) if supervisor_ids.any? }
scope :creator, ->(creator_id) { where(creator_id: creator_id) if creator_id.present? }
scope :contact_made, ->(contact_made) { where(contact_made: contact_made) if args[:contact_made] }
...
```

```
# app/models/case_contact_report.rb

def filtered_case_contacts(arg)
  CaseContact
    .supervisors(args[:supervisor_ids])
    .creator(args[:creator_id])
    .occurred_between(args[:start_date], args[:end_date])
    .contact_made(args[:made_contact])
    ...
end
```

With `inject`:

```
# app/models/case_contact.rb

def self.filtered_results(args)
    args.delete_if { |k, v| v.blank? }
    args.delete_if { |k, v| v.kind_of?(Array) and v.reject(&:empty?).empty? }
    args.inject(self.all) do |scope, (key, value)|
      case key
      when "supervisors_id"
        scope.where(supervisors: value)
      when "creator_id"
        scope.where(creator: value)
      when "has_transitioned"
        scope.joins(:casa_case).where("casa_cases.transition_aged_youth IN (?)", value)
      when 
        ...
      else
        scope
      end
    end
  end
end
```

Reply:
This will account for multiple filters :) it's building up one long query.

Reply:
Though I like your idea of moving these to scopes... 🤔

Reply:
Is that method getting called multiple times? I just wasn't seeing how it's building up to a single query if `contact` is getting redefined on each line...? But I am probably missing something!